### PR TITLE
Add profiler support

### PR DESCRIPTION
**Pull request by bwasti**
JSC's builtin profiler has the ability to dump profiles to a file or return JSON encoded runtime stack traces.

Three API's are exposed

```javascript
startSamplingProfiler(/*optional*/ dir_to_dump_profiles, /*optional*/ sample_rate_in_us);
const json = samplingProfilerStackTraces(); // get a running list of stack traces from the sampler
noInline(func); // tell the JIT compiler not to inline the named function - useful for getting better stack traces
```

### CODE REVIEW

**Comment by bwasti** [comment]
Example:

```javascript
import {startSamplingProfiler, samplingProfilerStackTraces, noInline} from "bun:jsc";

startSamplingProfiler(Bun.main + ".prof", 100);

function foo(a, b){
  let c = 0;
  for (let i = 0; i < 10000; ++i) {
    c += Math.sqrt(c) * a + b;
  }
  return c;
}
noInline(foo);

let tot = 0;
for (let i = 0; i < 10000; ++i) {
  tot += foo(i, Math.log(i));
}
console.log(tot);
//console.log(samplingProfilerStackTraces());
```

and the output:
```
Sampling rate: 100.000000 microseconds. Total samples: 5282
Top functions as <numSamples  'functionName#hash:sourceID'>
  5261    'foo#<nil>:2'
    11    '(module)#<nil>:2'
     8    'log#<nil>:4294967295'
     2    'sqrt#<nil>:4294967295'


Sampling rate: 100.000000 microseconds. Total samples: 5282

Tier breakdown:
-----------------------------------
LLInt:                     0  (0.000000%)
Baseline:                 75  (1.419917%)
DFG:                      81  (1.533510%)
FTL:                    5116  (96.857251%)
js builtin:                0  (0.000000%)
Wasm:                      0  (0.000000%)
Host:                      0  (0.000000%)
RegExp:                    0  (0.000000%)
C/C++:                     0  (0.000000%)
Unknown Executable:       10  (0.189322%)


Hottest bytecodes as <numSamples   'functionName#hash:JITType:bytecodeIndex'>
  3990    'foo#<nil>:FTL:bc#0'
  1126    'foo#<nil>:FTL:<nil>'
    73    'foo#<nil>:DFG:bc#23'
    46    'foo#<nil>:Baseline:bc#54'
    16    'foo#<nil>:Baseline:bc#23'
     8    'log#<nil>:None:<nil>'
     4    'foo#<nil>:DFG:<nil>'
     3    '(module)#<nil>:DFG:bc#175'
     3    '(module)#<nil>:Baseline:bc#175'
     2    'foo#<nil>:Baseline:bc#72'
     2    'foo#<nil>:Baseline:bc#60'
     2    'sqrt#<nil>:None:<nil>'
     1    '(module)#<nil>:DFG:bc#181'
     1    'foo#<nil>:Baseline:bc#84'
     1    '(module)#<nil>:Baseline:bc#156'
     1    '(module)#<nil>:Baseline:bc#135'
     1    'foo#<nil>:Baseline:bc#51'
     1    '(module)#<nil>:Baseline:bc#181'
     1    '(module)#<nil>:Baseline:bc#167'
```

**Comment by Jarred-Sumner** [comment]
I'll probably tweak the API a little but this looks great overall

I will update the WebKit build to export this this header and merge once building that is done (so that main branch doesn't break)

**Review by Jarred-Sumner — APPROVED** [review]

File: src/bun.js/bindings/BunJSCModule.cpp [hunk 1]
```diff
@@ -244,6 +247,62 @@ JSC_DEFINE_HOST_FUNCTION(functionCreateMemoryFootprint, (JSGlobalObject * global
...
+JSC_DEFINE_HOST_FUNCTION(functionStartSamplingProfiler, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    JSC::VM& vm = globalObject->vm();
+    JSC::SamplingProfiler& samplingProfiler = vm.ensureSamplingProfiler(WTF::Stopwatch::create());
+
+    JSC::JSValue directoryValue = callFrame->argument(0);
+    JSC::JSValue sampleValue = callFrame->argument(1);
+
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    if (directoryValue.isString()) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/BunJSCModule.cpp:266]: I recall `WTF::FileSystem` causing build issues on Ubuntu due to the usage of stdc++fs

but that might have gone away after upgrading to libc++20a

Comment (Jarred-Sumner) [reply src/bun.js/bindings/BunJSCModule.cpp:266]: but I can tweak this

Comment (bwasti) [reply src/bun.js/bindings/BunJSCModule.cpp:266]: any recommended APIs for FS? I was avoiding a C++ standard based on what I saw in the code.

**Comment by Jarred-Sumner** [comment]
planning to update webkit once https://github.com/WebKit/WebKit/pull/3508 is merged and then will merge this

**Pull request merged**